discojs/src/client/decentralized/base.ts

```diff
@@ -184,11 +149,25 @@ export abstract class Base extends ClientBase {
         throw new Error(`unknown protocol: ${this.url.protocol}`)
     }
     serverURL.pathname += `deai/${this.task.taskID}`
+
     this.server = await this.connectServer(serverURL)
 
-    await pauseUntil(() => this.connected)
+    const msg: messages.clientConnected = {
+      type: messages.type.clientConnected
+    }
+    this.server.send(msg)
+
+    const peerIdMsg = await waitMessage(this.server, messages.type.PeerID)
+    console.debug(peerIdMsg.id, 'got own id from server')
+
+    if (this.ID !== undefined) {
+      throw new Error('got ID from server but was already received')
+    }
+    this.ID = peerIdMsg.id
+    this.pool = PeerPool.init(peerIdMsg.id)
+    this.connected = true // Is this still needed?
 
-    console.debug(this.ID, 'server connected')
+    console.debug(this.ID, 'client connected to server')
   }
```

Comment:
Examples of "linear" flow with events:

1) Client send to the server "I am connected"
2) Waiting for server response with peerID (indirectly confirms the server is aware of us)
3) Setting everything up
4) Function returns, we can consider the connection is usable